### DISCUSSION TITLE
[V4] When using Deno and postcss it produces 22KB of unrelated css.

### DISCUSSION THREAD

**Discussion by tmikaeld** [tailwindlabs/tailwindcss#15082] [Help]
The same works in Tailwind V3, but here it produces 22KB of unrelated styles.

```javascript
import postcss from "npm:postcss";
import tailwindcss from "npm:@tailwindcss/postcss@next";

const tailwindBaseConfig = {
  content: [{ raw: '<div class="text-red-500"></div>', extension: "html" }],
  corePlugins: { preflight: false },
};

async function generateCSS(tailwindConfig) {
  try {
    const tailwindDirectives = `
      @tailwind components;
      @tailwind utilities;
    `;
    
    console.log('Config:', JSON.stringify(tailwindConfig, null, 2));
    console.log('Directives:', tailwindDirectives);

    const result = await postcss([tailwindcss(tailwindConfig)])
      .process(tailwindDirectives, { from: undefined });
    return result.css;
  } catch (error) {
    console.error("Error generating CSS:", error);
    throw error;
  }
}

console.log('Starting CSS generation...');
const css = await generateCSS(tailwindBaseConfig);
console.log('\nFinal CSS Size:', Math.round(css.length/1000)+' KB');
```

**Comment by philipp-spiess** [comment]
@tmikaeld Hey! It's hard to help here without a full reproduction but one thing you can check is what are the folders inside your current working dir (if nothing is configured in your CSS otherwise, we will do automatic content detection based on your current working dir).

You can customize this behavior with this API: https://tailwindcss.com/docs/v4-beta#adding-content-sources

**Reply by tmikaeld** [reply]
Hello @philipp-spiess 👋 

The source directory is empty, it's only this above test script. I tried with the `source: false` flag but that didn't change anything.

**Reply by philipp-spiess** [reply]
Do you mind putting up a repro to check? 

Also not really sure what `source: false` refers to, what I suggest is a CSS config:

```css
@import "tailwindcss" source(none);
```

**Reply by tmikaeld** [reply]
I don't know how to apply that in Deno, since it's not working.

With flag, i meant:
```
const tailwindBaseConfig = {
  content: [{ raw: '<div class="text-red-500"></div>', extension: "html" }],
  corePlugins: { preflight: false },
  "source": false
};
```

**Reply by philipp-spiess** [reply]
Oh, I see, you're still using v3 API. In V4, the JS config file can no longer be configured this way. Instead, you want to change your code to modify the CSS file input:

```js
const tailwindDirectives = `
  @import "tailwindcss" source(none);
`;
```

However if the idea is to pass class names programmatically (like in your `tailwindBaseConfig`, we don't support this right now). What's your use case for this over scanning the files directly?

**Reply by tmikaeld** [reply]
We're using it programmatically to generate a redis cache, scanning directories is a no-go for our use case.

It was the auto-scanning that was the issue, I ran it locally inside the folder and I get empty css, same as the repro.

https://github.com/tmikaeld/tw4/actions/runs/11970929929/job/33374679887

**Reply by tmikaeld** [reply]
So will programmatic support be dropped altogether or just added later?

We just got 3.x working and really don't want to create a fork for our use case.

**Reply by tmikaeld** [reply]
When using source(none) as suggested.
https://github.com/tmikaeld/tw4/actions/runs/11971285724/job/33375777839

**Reply by philipp-spiess** [reply]
We'll look into it for sure. For now, is it possible you write the class names to a `.txt` file in a tmp directory and then point the source detection to this `.txt` file instead?

**Reply by tmikaeld** [reply]
We'll stay on V3.x until we can use it programmatically. Thanks for the quick replies and good luck with the V4 launch!

**Reply by laktek** [reply]
@philipp-spiess curious if there are any updates regarding the programmatically setting content? I'm building a static site generator with built-in Tailwind support. Not having the `content` option has been a blocker to upgrade from v3 to v4.
